### DISCUSSION TITLE
How is code highlighting supposed to work?

### DISCUSSION THREAD

**Discussion by sherrellbc** [facebook/lexical#5428] [General]
The documentation on this is rather sparse (none?).

https://lexical.dev/docs/api/modules/lexical_code
https://github.com/facebook/lexical/blob/main/packages/lexical-playground/src/plugins/CodeHighlightPlugin/index.ts

I've pieced together this example from various sources online, as well as scouring the code and what documentation does exist (it's mostly just an API reference without any explanation of how to use it).

https://codesandbox.io/p/sandbox/stoic-pasteur-j3czn4

With the above code I can see in developer tools that the editor's code appears to be tokenized as you would expect ...

<img width="862" alt="Screenshot 2023-12-27 at 1 22 44 PM" src="https://github.com/facebook/lexical/assets/8580880/1e15d935-25d9-41c2-b30c-5d411f890bf8">

but no colors are applied

<img width="921" alt="Screenshot 2023-12-27 at 1 27 27 PM" src="https://github.com/facebook/lexical/assets/8580880/346daa42-13e1-4818-97fa-7b6e7c634101">

Am I missing something or is this an issue specifically with `@lexical/markdown`? From looking at the code I can see that [prismjs](https://github.com/facebook/lexical/blob/main/packages/lexical-code/src/CodeHighlightNode.ts#L20-L33) is used internally. Though I am unable to get it to properly apply.

Can code highlightning be used with either `RichTextPlugin` and `PlainTextPlugin`? I've tried both.

Also, shouldn't `<code>` generally be wrapped in `<pre>`? Without that I'm not sure how to distinguish `inline` code from `block` code.

**Comment by 2wheeh** [comment]
### My guess:

In the lexical state, code block is CodeNode and inline code is TextNode.
lexical uses a `<code>` for both when creating DOM:
block:
https://github.com/facebook/lexical/blob/9d247acc42998b4de3ed079069584ebde701640b/packages/lexical-code/src/CodeNode.ts#L104-L113
inline:
https://github.com/facebook/lexical/blob/9d247acc42998b4de3ed079069584ebde701640b/packages/lexical/src/nodes/LexicalTextNode.ts#L99-L102

And I think code highlighting should be used with `RichTextPlugin` as playground: https://github.com/facebook/lexical/blob/9d247acc42998b4de3ed079069584ebde701640b/packages/lexical-playground/src/Editor.tsx#L155

**Comment by sherrellbc** [comment]
Thanks for the reply here, but as you can see from my linked code sandbox  (and the picture above) there is no `<pre>` wrapping `<code>`. Either those unit tests are no longer working or it is testing some different use-case than I have contrived here.

> In the lexical state, code block is CodeNode and inline code is TextNode.
That may be, but I was concerned moreso with applying CSS to inline vs block code with e.g.
```
pre code {} //block code
code {} //inline code
```
But without `<pre>` I'm not sure how to do this. Anyway, code should be wrapped in a pre and a workaround should not be necessary here.

Code highlighting seems to work over at the [Playground](https://playground.lexical.dev/), but _only_ if you first apply the code in markdown and then switch back. Typing the code out and then marking it as a code block from the rich editor's code button does not work (and in fact the formatting is different from the markdown->text path as well).

<img width="1129" alt="Screenshot 2023-12-27 at 11 08 57 PM" src="https://github.com/facebook/lexical/assets/8580880/74e517b3-8893-4c6d-97fe-e4c7cc28995c">

The playground also does _not_ emit `<pre>` wrapping `<code>`, but it does very clearly apply many more classes to the tokenized code blocks that likely have something to do with which colors are applied; this does not happen in my code sandbox example. Not sure what the differences are.

<img width="619" alt="Screenshot 2023-12-27 at 11 10 44 PM" src="https://github.com/facebook/lexical/assets/8580880/01f0e121-98b1-4931-8ad6-b0823c5fe21e">

**Reply by sherrellbc** [reply]
You can see from this code that it is also transforming the markdown in a similar way. Though it appears to assume that first node of the root node will be a code block itself with language set to markdown. It creates this code just below in the default case. This is the only major difference I can spot.

https://github.com/facebook/lexical/blob/9d247acc42998b4de3ed079069584ebde701640b/packages/lexical-playground/src/plugins/ActionsPlugin/index.tsx#L143-L154

`PLAYGROUND_TRANSFORMERS` here is a custom list that prefixes a few transformers to the same ones exported from `@lexical/markdown`.

**Reply by 2wheeh** [reply]
> but it does very clearly apply many more classes to the tokenized code blocks that likely have something to do with which colors are applied; this does not happen in my code sandbox example

In your code, you also need to pass `theme` for the inline code and code block:
https://github.com/facebook/lexical/blob/9d247acc42998b4de3ed079069584ebde701640b/packages/lexical-playground/src/themes/PlaygroundEditorTheme.ts#L13-L18

The DOM you are seeing on the DevTools is created by `createDOM` through lexical state which have elements like `CodeNode`, `TextNode`.
and while creating DOM, classes are added as you pass as `theme`.

**Comment by sherrellbc** [comment]
I see. I certainly missed adding the theme. Now I see the updated classes in each of the items of the deconstructed code block. I used the same theme as the playground, so it now seems quite similar.

And so now highlighting is working as expected. Many thanks for your help here. I have used other highlighting tooling in the past that was far more opinionated at color scheme and such, so it was lost on me that both the theme object + CSS is required to make it work here.
